tpke/src/lib.rs

```diff
@@ -492,16 +476,158 @@ mod tests {
+        let rng = &mut test_rng();
+        let shares_num = 16;
+        let threshold = shares_num * 2 / 3;
+        let msg: &[u8] = "abc".as_bytes();
+        let aad: &[u8] = "my-aad".as_bytes();
+
+        let (pubkey, _, contexts) =
+            setup_simple::<E>(threshold, shares_num, rng);
+        let ciphertext = encrypt::<_, E>(msg, aad, &pubkey, rng);
+
+        // Remove one participant from the contexts and all nested structures
+        let mut remaining_participants = contexts;
+        remaining_participants.pop();
+        for p in &mut remaining_participants {
+            p.public_decryption_contexts.pop();
+        }
+
+        // Refresh the share
+        let x_r = Fr::rand(rng);
+        let y_r = recover_share_at_point(
+            &remaining_participants,
+            threshold,
+            &x_r,
+            rng,
+        );
+        let recovered_key_share = PrivateKeyShare {
+            private_key_shares: vec![y_r.into_affine()],
+        };
+
+        // Creating decryption shares
+        let mut decryption_shares: Vec<_> = remaining_participants
+            .iter()
+            .map(|ctxt| {
+                make_decryption_share(&ctxt.private_key_share, &ciphertext)
+            })
+            .collect();
+        decryption_shares
+            .push(make_decryption_share(&recovered_key_share, &ciphertext));
+
+        // Creating a shared secret from remaining shares and the recovered one
+        let shares_x = &remaining_participants[0]
+            .public_decryption_contexts
+            .iter()
+            .map(|ctxt| ctxt.domain)
+            .collect::<Vec<_>>();
+        let lagrange = prepare_combine_simple::<E>(shares_x);
+
+        let shared_secret =
+            share_combine_simple::<E>(&decryption_shares, &lagrange);
+
+        let plaintext = checked_decrypt_with_shared_secret(
+            &ciphertext,
+            aad,
+            &shared_secret,
+        );
+        assert_eq!(plaintext, msg);
+    }
+
+    #[test]
+    fn simple_threshold_decryption_with_share_refresh() {
+        let rng = &mut test_rng();
+        let shares_num = 16;
+        let threshold = shares_num * 2 / 3;
+        let msg: &[u8] = "abc".as_bytes();
+        let aad: &[u8] = "my-aad".as_bytes();
+
+        let (pubkey, _, contexts) =
+            setup_simple::<E>(threshold, shares_num, rng);
+        let ciphertext = encrypt::<_, E>(msg, aad, &pubkey, rng);
+
+        // Refresh shares
+        let fresh_shares = refresh_shares::<E>(&contexts, threshold, rng);
+
+        // Creating decryption shares
+        let decryption_shares: Vec<_> = fresh_shares
+            .iter()
+            .map(|private_share| {
+                let private_share = PrivateKeyShare {
+                    private_key_shares: vec![private_share.into_affine()],
+                };
+                make_decryption_share(&private_share, &ciphertext)
+            })
+            .collect();
+
```

Comment:
Same as above. After the share refreshing algorithm, the shared secret should remain the same.